### PR TITLE
Feature TT-3282 Add introspection resolver

### PR DESCRIPTION
**Pull request by spetrunin**
- Fix astparser bug with wrong Type.OfType value
- Add introspection datasource
- Attach introspection datasource to engine

### CODE REVIEW

File: pkg/engine/datasource/introspection_datasource/source.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	// fix corrupted input json
+	// TODO: remove when default argument value bug will be fixed
+	input = bytes.Replace(input, []byte(`"include_deprecated":}`), []byte(`"include_deprecated":false}`), 1)
+
+	var req introspectionInput
+	if err := json.Unmarshal(input, &req); err != nil {
+		return err
+	}
+
+	switch req.RequestType {
```
Comment (jensneuse) [inline pkg/engine/datasource/introspection_datasource/source.go:26]: What happens if a request contains multiple root fields with different Queries? e.g. requesting two type fields or type fields and schema?

File: pkg/ast/ast_root_operation_type_definition.go [hunk 2]
```diff
@@ -92,7 +93,9 @@ func (d *Document) ImportRootOperationTypeDefinition(name string, operationType
 	operationTypeDefinition := RootOperationTypeDefinition{
 		OperationType: operationType,
 		NamedType: Type{
-			Name: d.Input.AppendInputBytes(nameBytes),
+			Name:     d.Input.AppendInputBytes(nameBytes),
+			TypeKind: TypeKindNamed,
+			OfType:   -1,
```
Comment (pvormste) [inline pkg/ast/ast_root_operation_type_definition.go:98]: I still think we should reuse the constant we already have for this magic number :-)

File: pkg/astparser/parser.go [hunk 3]
```diff
@@ -280,6 +280,7 @@ func (p *Parser) parseRootOperationTypeDefinitionList(list *ast.RootOperationTyp
 				NamedType: ast.Type{
 					TypeKind: ast.TypeKindNamed,
 					Name:     namedType.Literal,
+					OfType:   -1,
```
Comment (pvormste) [inline pkg/astparser/parser.go:283]: Again magic number

File: pkg/engine/datasource/introspection_datasource/planner.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import (
+	"encoding/json"
+
+	"github.com/jensneuse/graphql-go-tools/pkg/engine/plan"
+	"github.com/jensneuse/graphql-go-tools/pkg/introspection"
+)
+
+type Planner struct {
+	introspectionData *introspection.Data
+	v                 *plan.Visitor
```
Comment (pvormste) [inline pkg/engine/datasource/introspection_datasource/planner.go:12]: I think I would have guessed `visitor` with `v` ... but why not make it really clear and write it out? :-)

**Review by pvormste — CHANGES_REQUESTED** [review]

**Pull request merged**